### PR TITLE
KEYCLOAK-2084 Client Registration CLI

### PR DESCRIPTION
**Pull request by mstruk**

### CODE REVIEW

**Comment by mstruk** [comment]
I don't understand the Travis failure. It says it can't find method clientRegistrationTrustedHost() on interface org.keycloak.admin.client.resource.RealmResource. That method has been added months ago.

**Comment by mstruk** [comment]
Merged #3322 is causing the following failure on keycloak server startup:

```
00:59:01,239 ERROR [org.jboss.msc.service.fail] (ServerService Thread Pool -- 54) MSC000001: Failed to start service jboss.undertow.deployment.default-server.default-host./auth: org.jboss.msc.service.StartException in service jboss.undertow.deployment.default-server.default-host./auth: java.lang.RuntimeException: RESTEASY003325: Failed to construct public org.keycloak.services.resources.KeycloakApplication(javax.servlet.ServletContext,org.jboss.resteasy.core.Dispatcher)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentService$1.run(UndertowDeploymentService.java:85)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
	at org.jboss.threads.JBossThread.run(JBossThread.java:320)
Caused by: java.lang.RuntimeException: RESTEASY003325: Failed to construct public org.keycloak.services.resources.KeycloakApplication(javax.servlet.ServletContext,org.jboss.resteasy.core.Dispatcher)
	at org.jboss.resteasy.core.ConstructorInjectorImpl.construct(ConstructorInjectorImpl.java:162)
	at org.jboss.resteasy.spi.ResteasyProviderFactory.createProviderInstance(ResteasyProviderFactory.java:2209)
	at org.jboss.resteasy.spi.ResteasyDeployment.createApplication(ResteasyDeployment.java:299)
	at org.jboss.resteasy.spi.ResteasyDeployment.start(ResteasyDeployment.java:240)
	at org.jboss.resteasy.plugins.server.servlet.ServletContainerDispatcher.init(ServletContainerDispatcher.java:113)
	at org.jboss.resteasy.plugins.server.servlet.HttpServletDispatcher.init(HttpServletDispatcher.java:36)
	at io.undertow.servlet.core.LifecyleInterceptorInvocation.proceed(LifecyleInterceptorInvocation.java:117)
	at org.wildfly.extension.undertow.security.RunAsLifecycleInterceptor.init(RunAsLifecycleInterceptor.java:78)
	at io.undertow.servlet.core.LifecyleInterceptorInvocation.proceed(LifecyleInterceptorInvocation.java:103)
	at io.undertow.servlet.core.ManagedServlet$DefaultInstanceStrategy.start(ManagedServlet.java:231)
	at io.undertow.servlet.core.ManagedServlet.createServlet(ManagedServlet.java:132)
	at io.undertow.servlet.core.DeploymentManagerImpl.start(DeploymentManagerImpl.java:526)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentService.startContext(UndertowDeploymentService.java:101)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentService$1.run(UndertowDeploymentService.java:82)
	... 6 more
Caused by: java.lang.RuntimeException: Invalid provider type 'org.keycloak.services.clientregistration.policy.ClientRegistrationPolicy'
	at org.keycloak.models.utils.ComponentUtil.getComponentFactory(ComponentUtil.java:54)
	at org.keycloak.models.jpa.RealmAdapter.addComponentModel(RealmAdapter.java:2039)
	at org.keycloak.services.clientregistration.policy.DefaultClientRegistrationPolicies.addAnonymousPolicies(DefaultClientRegistrationPolicies.java:85)
	at org.keycloak.services.clientregistration.policy.DefaultClientRegistrationPolicies.addDefaultPolicies(DefaultClientRegistrationPolicies.java:63)
	at org.keycloak.services.managers.RealmManager.setupClientRegistrations(RealmManager.java:606)
	at org.keycloak.services.managers.RealmManager.createRealm(RealmManager.java:124)
	at org.keycloak.services.managers.ApplianceBootstrap.createMasterRealm(ApplianceBootstrap.java:70)
	at org.keycloak.services.resources.KeycloakApplication.migrateAndBootstrap(KeycloakApplication.java:209)
	at org.keycloak.services.resources.KeycloakApplication$1.run(KeycloakApplication.java:137)
	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:231)
	at org.keycloak.services.resources.KeycloakApplication.<init>(KeycloakApplication.java:128)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
	at org.jboss.resteasy.core.ConstructorInjectorImpl.construct(ConstructorInjectorImpl.java:150)
	... 19 more
```

**Comment by mposolda** [comment]
@mstruk I've fixed the "Invalid provider type" in latest master. Regarding the removed clientRegistrationTrustedHost() , the trustedHosts were removed in latest master and replaced with dynamic client registration policies, which are more flexible.

**Comment by mstruk** [comment]
Any pointers on how to achieve the same through a new API?

https://github.com/mstruk/keycloak/blob/cli-reg/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/cli/registration/AbstractCliTest.java#L387-L397

**Comment by mposolda** [comment]
Yes, for example https://github.com/keycloak/keycloak/blob/master/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/client/ClientRegistrationPoliciesTest.java#L545-L550

File: integration/client-cli/client-registration-cli/README-windows.md [hunk 1]
```diff
@@ -0,0 +1,197 @@
+Keycloak Client Registration CLI
```
Comment (stianst) [inline integration/client-cli/client-registration-cli/README-windows.md]: Please remove this file. We don't want documentation in multiple places and this file will just end up being out of date.

File: integration/client-cli/client-registration-cli/README.md [hunk 2]
```diff
@@ -0,0 +1,201 @@
+Keycloak Client Registration CLI
```
Comment (stianst) [inline integration/client-cli/client-registration-cli/README.md]: Please remove this file. We don't want documentation in multiple places and this file will just end up being out of date.

File: integration/client-cli/client-registration-cli/pom.xml [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+            <artifactId>junit</artifactId>
+        </dependency>
+    </dependencies>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-shade-plugin</artifactId>
+                <version>2.4.3</version>
```
Comment (stianst) [inline integration/client-cli/client-registration-cli/pom.xml]: Move version declaration to pluginManagement in keycloak parent and use a property for the version

File: integration/client-cli/client-registration-cli/pom.xml [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+    <artifactId>keycloak-client-registration-cli</artifactId>
+    <name>Keycloak Client Registration CLI</name>
+    <description/>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.jboss.aesh</groupId>
+            <artifactId>aesh</artifactId>
+            <version>0.66.10</version>
```
Comment (stianst) [inline integration/client-cli/client-registration-cli/pom.xml]: Add version to dependencyManagement in keycloak parent and also use a property for it

Comment (mstruk) [inline integration/client-cli/client-registration-cli/pom.xml]: That doesn't work currently as it clashes with https://github.com/keycloak/keycloak/blob/2.2.1.Final/wildfly/adduser/src/main/java/org/keycloak/wildfly/adduser/AddUser.java which needs to be compatible with Wildfly's version of Aesh, whereas registration client has no dependency on Wildfly, and thus is technically not restricted in which version to use. Downgrading reg-cli to 0.66.4 doesn't work, also reg-cli will drive further Aesh improvements, and we need to be able to benefit from them. I think here we should keep reg-cli on a separate version.

Comment (stianst) [inline integration/client-cli/client-registration-cli/pom.xml]: OK

File: integration/client-cli/client-registration-cli/README.md (hunk 2 above)
Comment (mstruk) [inline integration/client-cli/client-registration-cli/README.md]: There is some content in there that is of use to QE team. Also, people may find at least pointers to documentation useful.

Maybe I can put testing notes into this PR's description.

**Pull request merged**